app/Services/Account/Subscription/ActivateLicenceKey.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace App\Services\Account\Subscription;
+
+use Exception;
+use Illuminate\Support\Str;
+use App\Services\BaseService;
+use App\Models\Account\Account;
+use Illuminate\Support\Facades\App;
+use Illuminate\Http\Client\Response;
+use Illuminate\Support\Facades\Http;
+use App\Exceptions\NoCustomerPortalSetException;
+use App\Exceptions\NoLicenceKeyEncryptionSetException;
+
+class ActivateLicenceKey extends BaseService
+{
+    private Account $account;
+    private array $data;
+    private Response $response;
+
+    /**
+     * Get the validation rules that apply to the service.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'account_id' => 'required|integer|exists:accounts,id',
+            'licence_key' => 'required|string:255',
+        ];
+    }
+
+    /**
+     * Check if the licence key given by the user is a valid licence key.
+     * If it is, activate the licence key and set the valid_until_at date.
+     *
+     * @param  array  $data
+     * @return void
+     */
+    public function execute(array $data): void
+    {
+        $this->validate($data);
+        $this->data = $data;
+        $this->account = Account::findOrFail($data['account_id']);
+
+        $this->validateEnvVariables();
+        $this->makeRequestToCustomerPortal();
+        $this->checkResponseCode();
+        $this->decodeAndStoreKey();
+    }
+
+    private function validateEnvVariables(): void
+    {
+        if (! config('monica.licence_key_encryption_key')) {
+            throw new NoLicenceKeyEncryptionSetException();
+        }
+
+        if (config('monica.customer_portal_url') == '') {
+            throw new NoCustomerPortalSetException();
+        }
+    }
+
+    private function makeRequestToCustomerPortal(): void
+    {
+        $url = config('monica.customer_portal_url').'/'.config('monica.customer_portal_secret_key').'/validate/'.$this->data['licence_key'];
```

Comment:
I'm not sure about how the customer portal is working yet, but it might be more secure to use OAuth2 here instead of a secret key in the url.

Reply:
OK, I think we should definitely use a client credential to implement this call: https://laravel.com/docs/9.x/passport#client-credentials-grant-tokens